Add spec for Signed Message (#5)
chaintope/tips: tip-0137.md

```
TIP: 0137
Type: Informational
Layer: Applications
Title: Generic Signed Message Format
Status: Draft
Author:  Hajime Yamaguchi <h_yamaguchi@chaintope.com>
Created: 2023-11-06
License: CC0
```

# Abstract

A standard for interoperable signed messages based on the Tapyrus Script format, either for proving fund availability, or committing to a message as the intended recipient of funds sent to the invoice address.

# Motivation

Since Tapyrus private keys can not only be used to sign Tapyrus transactions, but also any other message, it has become customary to use them to sign various messages for differing purposes. Some applications of signing messages with a Tapyrus private key are as follows: proof of funds for collateral, credit worthiness, entrance to events, airdrops, audits as well as other applications. 

> This TIP proposes a standard signature format using private key associated with utxo.

Reviewer: ```
This TIP proposes a standard signature format using a private key associated with a UTXO.
```

でしょうか？UTXO としているのは unspent なアウトプットである必要があるという意味でしょうか？
UTXO 出ないといけないと明確に記述しているところがなさそうなので、そういう仕様がなければ a transaction output の方がよさそうだと思いました。

Author: 「unspent である必要があるか？」ということでしょうか？
という意味であれば、ないです。記述を修正します。

Reviewer: わかりにくくてすいません。

>でしょうか？

の意図は`private key` のところは `a private key` でしょうか？それとも `private keys` でしょうか？ということでした。
マルチシグは念頭にないという話だったので、`a private key` だと思いますが。

As a similar specification, Bitcoin defines BIP137[^1] and BIP322[^2] as methods of signing messages using private keys, but this specification uses JWS as the message format.

# Specification

## Data Format

> Signed Message is a JSON Web Token (JWT) defined by RFC 7519[^3]

Reviewer: ```suggestion
A signed message is formatted JSON Web Token (JWT) defined by RFC 7519[^3].
```

固有名詞ではないので冠詞がいるかと思います。

> A message is represented as a string encoded in Base64url encoding and consists of three parts: Header, Payload, and Signature.

Reviewer: ```suggestion
The message is represented as a string encoded in Base64url encoding and consists of three parts: Header, Payload, and Signature.
```

### Header

> Header is a JSON Object which has the same structure as the JOSE Header defined in JWS(RFC-7515)[^4]

Reviewer: ```suggestion
The header is a JSON object which has the same structure as the JOSE Header defined in JWS(RFC-7515)[^4].
```

`JSON object` のように o は小文字にすることが一般的なようです。

> Header has the following 2 fields.

Reviewer: ```suggestion
The header has the following two fields.
```

* typ (OPTIONAL)
  * The "typ" (type) Header Parameter is used by JWT applications to declare the media type of this JWT. This specification supports only "JWT"
* alg (OPTIONAL)
  * The "alg" (algorithm) Header Parameter identifies the cryptographic algorithm used to secure the JWS. This specification supports only "ES256K"

> ### Payload

Reviewer: ペイロードにセットする txid, index に対応するアウトプットについて説明した上で、各クレームの説明を行った方が良いと思います。
クレームの説明中で `the Tapyrus transaction` とありますが、このトランザクションが何を指すのかが示されていないので。

また、trandsaction や scriptPubkey のフォーマットについてドキュメント全体を通じて言及されていませんが、特に縛りは無いですか？P2PKH, CP2PKH じゃないとだめとか。

Author: [4cabd38](https://github.com/chaintope/tips/pull/5/commits/4cabd38876898aec6759db022a5e3998cdff9400)
でサポートするアドレスのタイプについて追記してます。

Author: コミット漏れが有りました。9d1992ab91f12592c44861ce63de17beff1478dd で再度修正してます

Reviewer: > ペイロードにセットする txid, index に対応するアウトプットについて説明した上で、各クレームの説明を行った方が良いと思います。 

`#specification` の冒頭に以下のテキストを入れるのはどうでしょうか？

```
本仕様は、ある Tapyrus アウトプットに含まれる資金を移動させることができる秘密鍵を、あるメッセージの発信者が所有していることを示すことができるようにするものです。
ここではまずメッセージへの署名に用いる秘密鍵と、Tapyrus アウトプットの関係について説明します。

初めに、ある Tapyrus アウトプットがあります。そのアウトプットは ScriptPubkey に P2PKH または CP2PKH フォーマットのスクリプトをもっていなければいけません。
また、このアウトプットはメッセージの発信者がアンロックできるものでないといけません。
その発信者はそのアウトプットのアンロックに使用できる秘密鍵を用いてメッセージに署名します。

送信される署名済みメッセージは、その Tapyrus アウトプットを識別するための情報(TXID, アウトプットのindex）を含まなければいけません。
```

```
This specification aims to demonstrate that the sender of a message holds the private key. 
This key is capable of moving funds in a specific Tapyrus output. 
We will first clarify the relationship between the private key for message signing and the Tapyrus output.

There is a Tapyrus output under consideration. 
This output must have a script in ScriptPubkey. 
The script should be in either P2PKH (Pay to Public Key Hash) or CP2PKH (Colored Pay to Public Key Hash) format. 
Additionally, the output must be unlockable by the message sender.

The sender uses the private key, which can unlock the output, to sign the message.
The signed message sent must include the TXID and the index of the output to identify the relevant Tapyrus output.
```

Author: 729b917b1d97b49fc17143ca880d00c58be11dac でSpecificationの章にそのまま追加してます。

> Payload is a JSON Object called JWT Claims Set defined by JWT(RFC7519)

Reviewer: ```suggestion
The payload is a JSON object called JWT Claims Set defined by JWT(RFC7519).
```

> Payload has the following fields.

Reviewer: ```suggestion
The payload has the following fields.
```

* txid (REQUIRED)

>   * The "txid" (transaction ID) claim provides a identifier for the Tapyrus transaction. The "txid" value is a 64-characters hexadecimal string

Reviewer: ```suggestion
  * The "txid" (transaction ID) claim provides a identifier for the Tapyrus transaction. The "txid" value must be a 64-characters hexadecimal string.
```

* index (REQUIERD)

>   * The "index" claim indicates output index in the transaction which is owned by the signer. The "index" is a non-negative integer.

Reviewer: ```suggestion
  * The "index" claim indicates output index in the transaction which is owned by the signer. The "index" must be a non-negative integer.
```

* colorId (OPTIONAL)

>   * The "colorId" claim indicates a color identifier of the Utxo. The "colorId" is a 66-characters hexadecimal string

Reviewer: ```suggestion
  * The "colorId" claim indicates a color identifier of the UTXO. The "colorId" must be a 66-characters hexadecimal string
```

* value (REQUIRED)

>   * The "value" claim indicates the value of the Utxo. The "value" is a non-negative integer.

Reviewer: ```suggestion
  * The "value" claim indicates the value of the UTXO. The "value" must be a non-negative integer.
```

### Signature

The method of calculating the signature is defined in JWS.
ASCII(BASE64URL(UTF8(Header)) || '.' || BASE64URL(Payload)) is calculated with the HMAC SHA-256 algorithm and the result is base64url encoded.

### Example

The following is an example of Signed Message:

#### Header

The following is an example of a JSON object representing a header before it is encoded in Base64url:

```
{
  "alg": "ES256K",
  "typ": "JWT"
}
```

#### Payload

```
{
  txid: "f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16",
  index: 1,
  colorId: "c3ec2fd806701a3f55808cbec3922c38dafaa3070c48c803e9043ee3642c660b46",
  value: 1,
  scriptPubkey: "21c3ec2fd806701a3f55808cbec3922c38dafaa3070c48c803e9043ee3642c660b46bc76a91446c2fbfbecc99a63148fa076de58cf29b0bcf0b088ac",
  address: "22VdQ5VjWcF9zgsnPQodFBS1PBQPaAQEXSofkyMv2D9zV1MLp3JfScV6TMVaUQ42xeTfjieWssAaefMd",
  data: "00000000000000000000000000000000"
}
```

#### Signature(Encoded in Base64url format)

```
AdwMgeerwtHoh-l192l60hp9wAHZFVJbLfD_UxMi70cwnZOYaRI1bKPWROc-mZZqwqT2SI-KGDKB34XO0aw_7XdtAG8GaSwFKdCAPZgoXD2YBJZCPEX3xKpRwcdOO8KpEHwJjyqOgzDO7iKvU8vcnwNrmxYbSW9ERBXukOXolLzeO_Jn
```

## Verification

> A validator is given as input a JWT and public key, and outputs one of two states

Reviewer: ```suggestion
A validator is given a JWT and a public key as inputs, then outputs one of the two states below.
```

* `valid` indicates that the signature is valid
* `invalid` means that some check failed

Validation consists of the following steps:

>   1. Decode JWT form `signedMessage` as the 3 parts: `header`, `payload`, and `signature`

Reviewer: ```suggestion
  1. Decode JWT form a `signedMessage` as the 3 parts: `header`, `payload`, and `signature`
```

>   2. Confirm `header` satisfies the conditions.

Reviewer: ```suggestion
  2. Confirm the `header` satisfies the conditions.
```

       1. `typ` is `JWT`
       2. `alg` is `ES256K`

>   3. Confirm `payload` satisfies the conditions.

Reviewer: ```suggestion
  3. Confirm the `payload` satisfies the conditions.
```

        1. `txid` is a valid transaction id as 64-character hexdecimal string.
        2. `index` is non-negative integer less than 2^32
        3. `colorId` is a valid color id as 66-character hexdecimal string.
        4. `value` is non-negative integer less than 2^64
        5. `scriptPubkey` is a valid script as hexdecimal string.
        6. `address` is a Base58 string.
        7. `data` is a hexdecimal string whose length is less than xxxx.

>   4. Validate the JWS Signature(`signature`) against the JWS Signing Input ASCII(BASE64URL(`header`) || '.' || BASE64URL(`payload`)) in the manner defined for the algorithm `ES256K`.

Reviewer: 公開鍵の入手方法についての記述も必要かと思います。

Author: [4cabd38](https://github.com/chaintope/tips/pull/5/commits/4cabd38876898aec6759db022a5e3998cdff9400)
に公開鍵、秘密鍵についての記述を追加してます。「入手方法」についてはこの仕様の範囲外で良いかと思いましたが、どうしましょうか？

Author: コミット漏れが有りました。9d1992ab91f12592c44861ce63de17beff1478dd で再度修正してます

Reviewer: > 「入手方法」についてはこの仕様の範囲外で良いかと思いましたが、どうしましょうか？

ヘッダーに入れるのではだめですか？`jwk` 要素が使えるのではと思います。 https://tex2e.github.io/rfc-translater/html/rfc7517.html

Author: 70edcc58a107daf68b335604e1c98233bba1e9a2 でjwkをヘッダに追加してます。

>   5. Validate `payload` on Tapyrus Blockchain

Reviewer: ```suggestion
  5. Validate the `payload` on the Tapyrus blockchain
```

>      1. The transaction output indicated by `txid` and `index` exists on the Tapyrus Blockchain.

Reviewer: ```suggestion
     1. The transaction output indicated by the `txid` and `index` exists on the Tapyrus blockchain.
```

>      2. `value`, `scriptPubkey`, `address` are equal to those in the transaction above:

Reviewer: ```suggestion
     2. The `value`, `scriptPubkey`, and `address` are equal to those in the transaction output mentioned above:
```

Reviewer: あと、 scriptPubkey が JWT の署名を行った秘密鍵でアンロックできることも検証する必要があるかと思いましたが、いかがですか？

Reviewer: https://github.com/chaintope/tips/pull/5/files#r1406997009 で同じ話題に言及したので、こちらは解決とします。

## Signing

> Signers who control an address A who wish to sign a message `m` act as follows:

Reviewer: ```suggestion
Signers who control an address A and wish to sign a message `m` act as follows:
```

ちょっと読みにくく感じたので２つ目の who を and にしました。

Signers と複数形なのはマルチシグアドレスを想定しているという理解であっていますか？

Reviewer: マルチシグは想定していないという話でした。

>   1. Encode `header` as BASE64URL(UTF8(`header`))

Reviewer: ```suggestion
  1. Encode a `header` as BASE64URL(UTF8(`header`))
```

>   2. Encode `payload` as BASE64URL(`payload`)

Reviewer: ```suggestion
  2. Encode a `payload` as BASE64URL(`payload`)
```

>   3. Concatenating `header` and `payload` with a period ('.') character, 

Reviewer: ```suggestion
  3. Concatenate the `header` and the `payload` with a period ('.') character.
```

>   4. Pass two parameter `concat(header.payload)` and `private key` to ECDSA signing function. The result is the DER-format ECDSA digital signature.

Reviewer: ```suggestion
  4. Pass two parameters the `concat(header.payload)` and a `private key` to the ECDSA signing function. The result is the DER-format ECDSA digital signature.
```

>   5. Encode the signature as BASE64URL(JWS Signature)

Reviewer: ```suggestion
  5. Encode the signature as BASE64URL(JWS Signature).
```

>   6. Concatenating these values in the order `header`.`payload`.`signature` with period ('.') characters

Reviewer: ```suggestion
  6. Concatenate these values in the order `header`.`payload`.`signature` with period ('.') characters.
```

```
TIP: 0137
Type: Informational
Layer: Applications
Title: Generic Signed Message Format
Status: Draft
Author:  Hajime Yamaguchi <h_yamaguchi@chaintope.com>
Created: 2023-11-06
License: CC0
```

# Abstract

A standard for interoperable signed messages based on the Tapyrus Script format, either for proving fund availability, or committing to a message as the intended recipient of funds sent to the invoice address.

# Motivation

Since Tapyrus private keys can not only be used to sign Tapyrus transactions, but also any other message, it has become customary to use them to sign various messages for differing purposes. Some applications of signing messages with a Tapyrus private key are as follows: proof of funds for collateral, credit worthiness, entrance to events, airdrops, audits as well as other applications. 

> This TIP proposes a standard signature format using private key associated with a transaction output.

Reviewer: ```suggestion
This TIP proposes a standard signature format using a private key associated with a transaction output.
```

Author: 729b917b1d97b49fc17143ca880d00c58be11dac で修正してます。

## Data Format

A signed message is formatted JSON Web Token (JWT) defined by RFC 7519[^3].
The message is represented as a string encoded in Base64url encoding and consists of three parts: Header, Payload, and Signature.

### Header

The header is a JSON object which has the same structure as the JOSE Header defined in JWS(RFC-7515)[^4].
The header has the following two fields.

* typ (OPTIONAL)
  * The "typ" (type) Header Parameter is used by JWT applications to declare the media type of this JWT. This specification supports only "JWT"
* alg (OPTIONAL)
  * The "alg" (algorithm) Header Parameter identifies the cryptographic algorithm used to secure the JWS. This specification supports only "ES256K"

### Payload

The payload is a JSON object called JWT Claims Set defined by JWT(RFC7519).
The payload has the following fields.

* txid (REQUIRED)
  * The "txid" (transaction ID) claim provides a identifier for the Tapyrus transaction. The "txid" value must be a 64-characters hexadecimal string.
* index (REQUIERD)
  * The "index" claim indicates output index in the transaction which is owned by the signer. The "index" must be  a non-negative integer.
* colorId (OPTIONAL)

>   * The "colorId" claim indicates a color identifier of the transaction output. The "colorId" must be a 66-characters hexadecimal string. This specification supports only P2PKH and CP2PKH script pubkey.

Reviewer 2: `This specification supports only P2PKH and CP2PKH script pubkey.` は、scriptPubkeyに記載されているため、ここでは不要ではありませんか？

Author: a46b1bfb9ba9cc3b70c2d91c992b5d89e76946a7 で削除しました

> * value (REQUIRED)

Reviewer: value が必須になっている理由はなんでしょうか？

Reviewer: value, scriptPubkey, colorId、address は、バリデーション上は必要というわけではないけど、値が JWT に含まれると便利なので入れているという感じでしょうか？
また、それぞれvalue, scriptPubkey は必須、colorId, address はオプションと扱いが違う理由はなんでしょうか？

あと、それぞれの項目の想定されている用途などあれば知りたいです。

Author: トークンの保有を証明するというのが動機の１つとしてあると思うので、どのようなトークン(colorId)をどれだけ保有しているかの情報(value)があったほうが良いかと思いました、ただし、colorId, addressはscriptPubkeyがあればそこから取得できるためオプションとして、scriptPubkeyを必須にしてます。

Reviewer: 理解しました。ありがとうございます。

  * The "value" claim indicates the value of the transaction output. The "value" must be a non-negative integer.
* scriptPubkey (REQUIRED)
  * The "scriptPubkey" claim indicates the scriptPubkey script of the transaction output. The "scriptPubkey" must be a hexadecimal string. This specification supports only P2PKH and CP2PKH address.
* address(OPTIONAL)
  * The "address" claim indicates the address of the transaction output to prove ownership. The "address" must be a Base58 string.

> * data(OPTIONAL)

Reviewer 2: ```suggestion
* message(OPTIONAL)
```

Author: Fixed at https://github.com/chaintope/tips/commit/c6ca934c8d855894ea6364b42ee9abf3ed68cdf7

>   * Any message data to be signed. The "data" value must be a hexadecimal string.

Reviewer 2: ```suggestion
  * Any message to be signed. The "message" value must be a hexadecimal string.
```

Author: Fixed at https://github.com/chaintope/tips/pull/5/commits/c6ca934c8d855894ea6364b42ee9abf3ed68cdf7

  value: 1,
  scriptPubkey: "21c3ec2fd806701a3f55808cbec3922c38dafaa3070c48c803e9043ee3642c660b46bc76a91446c2fbfbecc99a63148fa076de58cf29b0bcf0b088ac",
  address: "22VdQ5VjWcF9zgsnPQodFBS1PBQPaAQEXSofkyMv2D9zV1MLp3JfScV6TMVaUQ42xeTfjieWssAaefMd",
  data: "00000000000000000000000000000000"
}
```

#### Signature(Encoded in Base64url format)

```
AdwMgeerwtHoh-l192l60hp9wAHZFVJbLfD_UxMi70cwnZOYaRI1bKPWROc-mZZqwqT2SI-KGDKB34XO0aw_7XdtAG8GaSwFKdCAPZgoXD2YBJZCPEX3xKpRwcdOO8KpEHwJjyqOgzDO7iKvU8vcnwNrmxYbSW9ERBXukOXolLzeO_Jn
```

## Verification

A validator is given a JWT and a public key as inputs, then outputs one of the two states below.

* `valid` indicates that the signature is valid
* `invalid` means that some check failed

Validation consists of the following steps:

  1. Decode JWT form a `signedMessage` as the 3 parts: `header`, `payload`, and `signature`.
  2. Confirm the `header` satisfies the conditions.
       1. `typ` is `JWT`
       2. `alg` is `ES256K`
  3. Confirm the `payload` satisfies the conditions.
        1. `txid` is a valid transaction id as 64-character hexdecimal string.
        2. `index` is non-negative integer less than 2^32.

>         3. `colorId` is a valid color id as 66-character hexdecimal string.

Reviewer 2: `colorId`が`scriptPubkey`のcolorIdと同じかチェックする必要がありませんか？

Author: https://github.com/chaintope/tips/pull/5/commits/db5a7285935808ed16f7475c87ed9270ad0dbc8f で修正してます。

        4. `value` is non-negative integer less than 2^64.
        5. `scriptPubkey` is a valid script as hexdecimal string.
        6. `address` is a Base58 string.
        7. `data` is a hexdecimal string whose length is less than xxxx.
  4. Validate the JWS Signature(`signature`) against the JWS Signing Input ASCII(BASE64URL(`header`) || '.' || BASE64URL(`payload`)) in the manner defined for the algorithm `ES256K`.
  5. Validate the `payload` on the Tapyrus blockchain
     1. The transaction output indicated by the `txid` and `index` exists on the Tapyrus blockchain.
     2. The `value`, `scriptPubkey`, and `address` are equal to those in the transaction output mentioned above.

This validation process requires a `public key` provided by the signer.

> This public key must be identical to the public key used to generate the output address of the transaction

Reviewer: ```suggestion
This public key must be identical to the public key used to generate the output address of the transaction.
```

> This public key must be identical to the public key used to generate the output address of the transaction

Reviewer: この内容もバリデーションのステップの中に入れるべきでは？

Reviewer: https://github.com/chaintope/tips/pull/5#discussion_r1411567100 により具体的に書きましたので、こちらのスレッドは resolve とします。

> This public key must be identical to the public key used to generate the output address of the transaction

Reviewer 2: BitcoinやEthereumで使われているのと同様に署名データから公開鍵を復元するのでもいいかもしれません。

Author: https://github.com/chaintope/tips/commit/70edcc58a107daf68b335604e1c98233bba1e9a2 で公開鍵をjwk形式でヘッダに追加してます。

Only public keys in compressed form (33 bytes) may be used in this specification.

## Signing

> Signers who control an address A and wish to sign a message `m` act as follows:

Reviewer: ```suggestion
A signer who control an address A and wish to sign a message `m` act as follows:
```

Signers だと複数人のサイナーが1つのアドレスを管理しているように読めてしまうので、`A signer` の方が良いと思います

  value: 1,
  scriptPubkey: "21c3ec2fd806701a3f55808cbec3922c38dafaa3070c48c803e9043ee3642c660b46bc76a91446c2fbfbecc99a63148fa076de58cf29b0bcf0b088ac",
  address: "22VdQ5VjWcF9zgsnPQodFBS1PBQPaAQEXSofkyMv2D9zV1MLp3JfScV6TMVaUQ42xeTfjieWssAaefMd",
  message: "00000000000000000000000000000000"
}
```

#### Signature(Encoded in Base64url format)

```
AdwMgeerwtHoh-l192l60hp9wAHZFVJbLfD_UxMi70cwnZOYaRI1bKPWROc-mZZqwqT2SI-KGDKB34XO0aw_7XdtAG8GaSwFKdCAPZgoXD2YBJZCPEX3xKpRwcdOO8KpEHwJjyqOgzDO7iKvU8vcnwNrmxYbSW9ERBXukOXolLzeO_Jn
```

## Verification

A validator is given a JWT, then outputs one of the two states below.

* `valid` indicates that the signature is valid
* `invalid` means that some check failed

Validation consists of the following steps:

  1. Decode JWT form a `signedMessage` as the 3 parts: `header`, `payload`, and `signature`.
  2. Confirm the `header` satisfies the conditions.
       1. `typ` is `JWT`
       2. `alg` is `ES256K`
       3. `jwk`.`keys` is an array that has only element, and the element satisfieds the conditions:
          1. `kty` is `EC`
          2. `crv` is `P-256K`
          3. `use` is `sig`
          4. `alg` is `ES256K`

>           5. `x` and `y` can be recovered as points on an elliptic curve, and the P2PKH or CP2PKH address derived from the the point is the same as the `address` in the `payload`

Reviewer: ```suggestion
          5. The `x` and `y` coordinates can be recovered as a point on the elliptic curve specified by the `crv` element, which in this case is the secp256k1 curve. And the P2PKH or CP2PKH address derived from the the point is the same as the `address` in the `payload`
```

細かいですが、楕円曲線についての指定も明示した方が良いと思いました。

>           5. `x` and `y` can be recovered as points on an elliptic curve, and the P2PKH or CP2PKH address derived from the the point is the same as the `address` in the `payload`

Reviewer: > and the P2PKH or CP2PKH address derived from the the point is the same as the `address` in the `payload`

はバリデーションのステップとしては不要だと思いました。
`address` はオプションの項目なので、必ずしも実施できることではないためです。

もしくは、公開鍵がアウトプットをロックしている鍵と同じかを検証するという意図であれば、scriptPubkey を使う方が良いと思います。
例えば、ステップ3の後に以下を加えるのはいかがでしょうか？

```
4. Confirm the `jwk` key corresponds to the public key hash in the `scriptPubkey` in the `payload`.
     1. Compress Public Key: Start by constructing a compressed public key binary from the `x` and `y` coordinates found in the `jwk`.
     2. Apply Hash Functions: Subject the binary representation of the compressed public key to a series of hash functions, specifically SHA-256 followed by RIPEMD-160. This sequence is denoted as SHA-256(RIPEMD-160(public key binary)).
     3. Verify Public Key Hash: Ensure that the resultant hash value from step 2 matches the public key hash extracted from the `scriptPubkey`.
```

Author: https://github.com/chaintope/tips/pull/5/commits/4d32e7c865c65de8504c517c214d8c7547d8b119 で追記してます。

* txid (REQUIRED)
  * The "txid" (transaction ID) claim provides a identifier for the Tapyrus transaction. The "txid" value must be a 64-characters hexadecimal string.
* index (REQUIERD)
  * The "index" claim indicates output index in the transaction which is owned by the signer. The "index" must be  a non-negative integer.
* colorId (OPTIONAL)
  * The "colorId" claim indicates a color identifier of the transaction output. The "colorId" must be a 66-characters hexadecimal string.
* value (REQUIRED)
  * The "value" claim indicates the value of the transaction output. The "value" must be a non-negative integer.
* scriptPubkey (REQUIRED)
  * The "scriptPubkey" claim indicates the scriptPubkey script of the transaction output. The "scriptPubkey" must be a hexadecimal string. This specification supports only P2PKH and CP2PKH script pubkey.
* address(OPTIONAL)
  * The "address" claim indicates the address of the transaction output to prove ownership. The "address" must be a Base58 string.
* message(OPTIONAL)
  * Any message to be signed. The "message" value must be a hexadecimal string.

### Signature

The method of calculating the signature is defined in JWS.

> ASCII(BASE64URL(UTF8(Header)) || '.' || BASE64URL(Payload)) is calculated with the HMAC SHA-256 algorithm and the result is base64url encoded.

Reviewer: Header は UTF8 とエンコードを指定していますが、Payload にはありませんが不要ですか？

Author: https://datatracker.ietf.org/doc/html/rfc7515#section-3.1
の説明をそのまま取ってきてますが、Payload自体がUTF-8でエンコーディングされてること前提のような気がします。

Reviewer: JWS の仕様を引用している形ですね。了解です。

Example JWS のセクションでペイロードは UTF-8表現って書いてますね。
Header は `UTF8()` で囲ってペイロードはそうでないのはなぜなんでしょうね。Example JWS では UTF-8 を使っているけど仕様としては決めてないので、任意のエンコードを使えるようにしているってことですかね。

Author: 説明を修正して、JWS(rfc7515)由来であること明記しました。
b8ec2d7

  3. Concatenate the `header` and the `payload` with a period ('.') character.
  4. Pass two parameters the `concat(header.payload)` and a `private key` to the ECDSA signing function. The result is the DER-format ECDSA digital signature.
  5. Encode the signature as BASE64URL(JWS Signature).
  6. Concatenate these values in the order `header`.`payload`.`signature` with period ('.') characters.

This signing process requires a `private key` consisting of randomly generated numbers. 
The private key used in this process must be identical to the private key used to unlock the transaction output.
In Tapyrus, the private key is a single unsigned 256-bit integer (32 bytes).

# Reference implementation

- [Tapyrusrb](https://github.com/chaintope/tapyrusrb/lib/tapyrus/tip0137.rb)

# References

- [^1] [BIP 0137](https://github.com/bitcoin/bips/blob/master/bip-0137.mediawiki)
- [^2] [BIP 0322](https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki)
- [^3] [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519)
- [^4] [RFC 7515](https://datatracker.ietf.org/doc/html/rfc7515)
- [^5] [RFC 7517](https://datatracker.ietf.org/doc/html/rfc7517)

> # Test vectors

Reviewer: 以下のケースについてもテストベクターがあった方が良いと思いました。

* script pubkey が p2pkh の場合
* 検証に失敗するケースをいくつか
     * 署名の検証に失敗するケース
     * 署名の検証に使用する公開鍵が script pubkey と一致しないケース
     * ペイロードの内容が TX の内容と一致しないケース
     * ほかにあれば

Author: [3da92fc](https://github.com/chaintope/tips/pull/5/commits/3da92fc4af1f9ea394f3c0b7530dd930d301acd1)

- scriptがp2pkhの場合
- 署名検証に失敗する場合
- colorIdが不正の場合
- p2sh/cp2shの場合
についてパターンを追加してます。

Reviewer: > * 署名の検証に使用する公開鍵が script pubkey と一致しないケース
> * ペイロードの内容が TX の内容と一致しないケース

こちらのケースはいかがですか？
後者はともかく前者のほうはあった方が良いかと思いました。

Author: > 署名の検証に使用する公開鍵が script pubkey と一致しないケース

このケースで追加してる認識です。

https://github.com/chaintope/tips/pull/5/files#diff-4aba63856c8e0ff3b5638dfe72c898105e9899e42b7712104bb17883979ed02fR257-R284

Fields to be signed is the followings:

> - txid: 0101010101010101010101010101010101010101010101010101010101010101

Reviewer: TXのペイロードも付与した方が良いと思います。実際の処理では TX のペイロードと JWT のペイロードを比較して検証するので。
さすがにオンチェーンの TX までは用意しないでいいとおもいますが。

Author: [3da92fc](https://github.com/chaintope/tips/pull/5/commits/3da92fc4af1f9ea394f3c0b7530dd930d301acd1)
でTx全体も追記してます。

The public key(hexadecimal string): 034f355bdcb7cc0af728ef3cceb9615d90684bb5b2ca5f859ab0f0b704075871aa

The signed message:

```
eyJ0eXAiOiJKV1QiLCJqd2siOnsia2V5cyI6W3sia3R5IjoiRUMiLCJjcnYiOiJQLTI1NksiLCJ4IjoiVHpWYjNMZk1DdmNvN3p6T3VXRmRrR2hMdGJMS1g0V2FzUEMzQkFkWWNhbyIsInkiOiJPRnRyRzQ2dGdKeW1kRlRaYURfUEs2QTBWdGItTEVxLUt3ZnctOXV5OGNFIiwidXNlIjoic2lnIiwiYWxnIjoiRVMyNTZLIiwia2lkIjoiZTc1ZDRjYTI3Nzg3MzdjMTQwNGY1MzRkMTQ2NWUyY2Q0NmJiMWY3MWRmODZjNjFkN2M5NDEzZTVjYTFhYTY5MCJ9XX0sImFsZyI6IkVTMjU2SyJ9.eyJ0eGlkIjoiMDEwMTAxMDEwMTAxMDEwMTAxMDEwMTAxMDEwMTAxMDEwMTAxMDEwMTAxMDEwMTAxMDEwMTAxMDEwMTAxMDEwMSIsImluZGV4IjoxLCJjb2xvcklkIjoiYzNlYzJmZDgwNjcwMWEzZjU1ODA4Y2JlYzM5MjJjMzhkYWZhYTMwNzBjNDhjODAzZTkwNDNlZTM2NDJjNjYwYjQ2IiwidmFsdWUiOjEsInNjcmlwdFB1YmtleSI6IjIxYzNlYzJmZDgwNjcwMWEzZjU1ODA4Y2JlYzM5MjJjMzhkYWZhYTMwNzBjNDhjODAzZTkwNDNlZTM2NDJjNjYwYjQ2YmM3NmE5MTRmYzcyNTBhMjExZGVkZGM3MGVlNWEyNzM4ZGU1ZjA3ODE3MzUxY2VmODhhYyIsImFkZHJlc3MiOiIyMlZkUTVWaldjRjl6Z3NuUFFvZEZCUzFQQlFQYUFRRVhTb2ZreU12MkQ5elYxTWROaGVhQXk3c3JvVGc1Mm13VzVhcE5oeFBxQjZYNFlSRyIsIm1lc3NhZ2UiOiIwMTAyMDMwNDA1MDYwNzA4MDkwYTBiMGMwZDBlMGYifQ.9X6wpnjwXRzzCcWiE2c3cKuuQVUrgxwGzlHHSixXyW_rMguqAnnE3SqzFbm00ZYOxqx00A9ExGM0GMqMnsbTiw

> ```

Reviewer: jwk の中に以下の値が入っているようですが、これはどのように生成される値ですか？また、必要な項目であれば仕様に入れる必要があると思います。
```
"kid": "e75d4ca2778737c1404f534d1465e2cd46bb1f71df86c61d7c9413e5ca1aa690"
```

Author: ruby-jwtを使用してjwkを生成すると自動で割り振られる数値です。JWKの仕様的には任意の項目です。

Reviewer: 公開鍵のハッシュ値が入るようですね。了解しました。

    * alg
      * The "alg" (algorithm) parameter identifies the algorithm intended for use with the key. This specification supports only "ES256K"

### Payload

The payload is a JSON object called JWT Claims Set defined by JWT(RFC7519).
The payload has the following fields.

* txid (REQUIRED)
  * The "txid" (transaction ID) claim provides a identifier for the Tapyrus transaction. The "txid" value must be a 64-characters hexadecimal string.
* index (REQUIERD)
  * The "index" claim indicates output index in the transaction which is owned by the signer. The "index" must be  a non-negative integer.
* colorId (OPTIONAL)
  * The "colorId" claim indicates a color identifier of the transaction output. The "colorId" must be a 66-characters hexadecimal string.
* value (REQUIRED)
  * The "value" claim indicates the value of the transaction output. The "value" must be a non-negative integer.
* scriptPubkey (REQUIRED)
  * The "scriptPubkey" claim indicates the scriptPubkey script of the transaction output. The "scriptPubkey" must be a hexadecimal string. This specification supports only P2PKH and CP2PKH script pubkey.
* address(OPTIONAL)

>   * The "address" claim indicates the address of the transaction output to prove ownership. The "address" must be a Base58 string.

Reviewer 2: scriptPubkeyから導出されるアドレスと一致するか検証する必要がありませんか？

Author: ad7ecc0548951d137808b9a12988c02e7a02ee35 で記述を追加してます。
